### ISSUE TITLE
Batch normalization 

### ISSUE THREAD

**Issue by neel-dey** [run-ai/runai#14]
Thanks for the easy to use gradient accumulation implementation.

I was wondering how batch normalization is handled here. If a model uses BN, are the running statistics updated per sample (inner iteration) or per gradient update (outer iteration)? What about the affine scale and shift terms?

**Comment by levosos** [comment]
Hi @neel-dey and thank you for reaching out!
This is a really good question.

As we covered in this Medium [article](https://towardsdatascience.com/how-to-break-gpu-memory-boundaries-even-with-large-batch-sizes-7a9c27a400ce):

> Batch normalization is being done separately on every mini-batch and not on the global batch

This is the reason why a GA version of a model that uses BN is not completely equivalent to non-GA version of it using the global batch size.

This means the statistics are updated per mini-batch of samples, every "inner" step.

Please share any further thoughts you'll have regarding this, and thanks again for reaching out!

**Comment by neel-dey** [comment]
Understood. Thank you :)
